envoy/config/filter/http/jwt_authn/v2/config.proto

```diff
@@ -0,0 +1,237 @@
+  // If true, the JWT is not removed in the request. Default value is false.
+  bool forward_jwt = 5;
+
+  // Two fields below define where to extract the JWT from an HTTP request.
+  //
+  // If no explicit location is specified, the following default locations are
+  // tried in order:
+  //
+  // 1. The Authorization header using the Bearer schema.
+  //    See `here <https://tools.ietf.org/html/rfc6750#section-2.1>`_
+  //    Example::
+  //
+  //       Authorization: Bearer <token>.
+  //
+  // 2. `access_token` query parameter.
+  //    See `this <https://tools.ietf.org/html/rfc6750#section-2.3>`_
+  //
+
+  // Multiple JWTs can be verified for a request. Each JWT has to be extracted
+  // from the locations its issuer specified or from the default locations.
+
+  // Specify the HTTP headers to extract JWT token.
+  //
+  // For examples, following config::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_headers:
+  //   - name: x-goog-iap-jwt-assertion
+  //
+  // can be used to extract token from header::
+  //
+  //   x-goog-iap-jwt-assertion: <JWT>.
+  //
+  repeated JwtHeader jwt_headers = 6;
+
+  // JWT is sent in a query parameter. `jwt_params` represents the
+  // query parameter names.
+  //
+  // For example, if config is::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_params:
+  //   - jwt_token
+  //
+  // The JWT format in query parameter is::
+  //
+  //    /path?jwt_token=<JWT>
+  //
+  repeated string jwt_params = 7;
+
+  // This field specifies the header name to forward a successfully verified
+  // JWT payload to the backend. The forwarded data is::
+  //
+  //    base64_encoded(jwt_payload_in_JSON)
+  //
+  // If it is not specified, the payload will not be forwarded.
+  // If multiple JWT payloads needed to be forwarded, distinct header names
+  // are required.
+  string jwt_payload_header = 8;
+}
+
+// This message specifies how to fetch JWKS from remote and how to cache it.
+message RemoteJwks {
+  // The HTTP URI to fetch the JWKS.
+  //
+  // Example:
+  //
+  // .. code-block:: yaml
+  //
+  //    http_uri:
+  //    - uri: https://www.googleapis.com/oauth2/v1/certs
+  //      cluster: jwt.www.googleapis.com|443
+  //
+  envoy.api.v2.core.HttpUri http_uri = 1;
+
+  // Duration after which the cached JWKS should be expired. If not
+  // specified, default cache duration is 5 minutes.
+  google.protobuf.Duration cache_duration = 2;
+}
+
+// This message specifies a header location to extract JWT token.
+message JwtHeader {
+  // The HTTP header name.
+  string name = 1 [(validate.rules).string.min_bytes = 1];
+
+  // The value prefix. The value format is "value_prefix<token>"
+  // For example, for "Authorization: Bearer <token>",
+  // value_prefix="Bearer " with a space at the end.
+  string value_prefix = 2;
+}
+
+// This is the Envoy HTTP filter config for JWT authentication.
+// [#not-implemented-hide:]
+message JwtAuthentication {
+  // List of JWT rules to valide.
+  repeated JwtRule jwt_rules = 1;
```

Comment:
Throughout this file, I feel many `jwt_` prefix and `_jwt` suffix are redundant as they are already in `Jwt*` messages, remove them?

Reply:
It is redundant,  but it is still better to have Jwt prefix.  It makes it easier to read. Otherwise, just "Rules" or "Authentication",  only the first glimpse, not exactly sure what Rule and Authentication it is referring to.

Reply:
I just talking about fields, not about the message names. here `rules` seems sufficient.

Reply:
Done.